### PR TITLE
Remove TensorFlow dependency, Fix fit batches and add history

### PR DESCRIPTION
**Pull request by Daniel-Dodd**
This PR removes TensorFlow (#69  ). Though the mini-batch cacheing bug in the function `fit_batches` (#96 ) can be fixed by JAX's host callback, as per a previous commit on this branch, it is desirable to remove this dependency.

Please check the type of change your PR introduces:

- [ X] Bugfix
- [ X] Feature
- [ X] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ X] Documentation content changes
- [ ] Other (please describe):

## Other information

### CODE REVIEW

File: gpjax/abstractions.py [hunk 1]
```diff
@@ -125,42 +130,56 @@ def fit_batches(
...
         log_rate (int, optional): How frequently the objective function's value should be printed. Defaults to 10.
     Returns:
-        tp.Dict: An optimised set of parameters.
+        tp.Tuple[tp.Dict, f64["n_iters"]]: A tuple comprising optimised parameters and training history respectively.
     """
 
     opt_state = optax_optim.init(params)
-    next_batch = train_data.get_batcher()
+
+    prng = convert_seed(seed)
```
Comment (thomaspinder) [inline gpjax/abstractions.py:156]: Is there a specific reason for this? Why can't we just require the user to supply a Jax PRNGKey?

File: gpjax/abstractions.py [hunk 2]
```diff
@@ -125,42 +130,56 @@ def fit_batches(
...
+        batch = Dataset(X=x[indicies], y=y[indicies])
+
         loss_val, loss_gradient = jax.value_and_grad(loss)(params, batch)
         updates, opt_state = optax_optim.update(loss_gradient, opt_state, params)
         params = optax.apply_updates(params, updates)
-        params_opt_state = params, opt_state
-        return params_opt_state, loss_val
 
-    (params, _), _ = jax.lax.scan(step, (params, opt_state), jnp.arange(n_iters))
+        prng, _ = jr.split(prng)
```
Comment (thomaspinder) [inline gpjax/abstractions.py:176]: Are you sure this is correct? Typically in Jax we'd split the key as `key, subkey = jr.split(key)` and then use subkey for stochastic ops. I believe this is because the return of `jr.split(key, 10) will first return a single key and the seond return is a list of 10 keys for use in a stochastic ops.

File: gpjax/types.py [hunk 3]
```diff
@@ -45,108 +39,6 @@ def out_dim(self) -> int:
...
-        self._make_tfd_if_none()
-        tfd_iter = iter(self._tf_dataset)
-
-        def next_batch() -> "Dataset":
-            X_batch, y_batch = next(tfd_iter)
-
-            return Dataset(X=X_batch.numpy(), y=y_batch.numpy())
-
-        return next_batch
-
```
Comment (thomaspinder) [inline gpjax/types.py:149]: This is very pleasing :)

File: gpjax/utils.py [hunk 4]
```diff
@@ -77,3 +79,14 @@ def array_to_dict(parameter_array) -> tp.Dict:
...
+
+def convert_seed(seed: tp.Union[int, PRNGKey]) -> PRNGKey:
+    """Ensure that seeds type."""
+
+    if isinstance(seed, int):
+        rng = jr.PRNGKey(seed)
+    else:  # key is of type PRNGKey
+        rng = seed
+
+    return rng
```
Comment (thomaspinder) [inline gpjax/utils.py:92]: I feel this may be superfluous?

**Review by thomaspinder — CHANGES_REQUESTED** [review]
This PR makes me very happy. I have a few comments that I'd like resolving, then we can merge.

File: gpjax/abstractions.py [hunk 5]
```diff
@@ -2,13 +2,17 @@
 
 import jax
 import jax.numpy as jnp
+import jax.random as jr
 import optax
+from chex import PRNGKey
```
Comment (thomaspinder) [inline gpjax/abstractions.py:7]: Is this intentional? Could we perhaps alias the import as `PRNGKeyType`

File: gpjax/abstractions.py (hunk 1 above)
Comment (Daniel-Dodd) [reply gpjax/abstractions.py:156]: The thinking behind this was to match how things are done in distrax. With fresh eyes, I would agree that this is superfluous. Lets remove it.

File: gpjax/abstractions.py (hunk 2 above)
Comment (Daniel-Dodd) [reply gpjax/abstractions.py:176]: Ooh I did not realise this. I will have to fix this. I think we might want to have the data batch loading code as a separate function so that we can unit test it.

File: gpjax/types.py (hunk 3 above)
Comment (Daniel-Dodd) [reply gpjax/types.py:149]: It was satisfying hitting backspace!

File: gpjax/utils.py (hunk 4 above)
Comment (Daniel-Dodd) [reply gpjax/utils.py:92]: Agreed. I'll remove this.

File: gpjax/abstractions.py (hunk 5 above)
Comment (Daniel-Dodd) [reply gpjax/abstractions.py:7]: Good spot! I did not give this much thought, I followed what was done in distrax (see e.g., https://github.com/deepmind/distrax/blob/615605e8701eb55f75854b9b6e13913a9d6d7716/distrax/_src/distributions/distribution.py). Your suggestion of `PRNGKeyType` makes more sense!

**Comment by Daniel-Dodd** [comment]
Thanks for a thorough review @thomaspinder . I'll get on with the changes. Then I can do the same for the natural gradients branch, and then that PR it ready to be reviewed too.

**Comment by Daniel-Dodd** [comment]
@thomaspinder I have resolved the changes.

I have put the batch selection and key updating code in to a separate function. Though less neat, it means we can ensure unit tests assert the correct batch size, dimensions, batch updates are used. Thoughts?

**Comment by Daniel-Dodd** [comment]
@thomaspinder have done a rebase to account for the #100 merge. I need to update the tests before you review.

**Comment by Daniel-Dodd** [comment]
@thomaspinder Ready for review.

File: gpjax/abstractions.py [hunk 6]
```diff
@@ -162,33 +161,44 @@ def fit_batches(
...
 
     @progress_bar_scan(n_iters, log_rate)
     def step(carry, _):
-        params, opt_state, prng = carry
-
-        indicies = jr.choice(prng, n, (batch_size,), replace=True)
+        params, opt_state, current_key = carry
 
-        batch = Dataset(X=x[indicies], y=y[indicies])
+        batch, new_key = get_batch_new_key(train_data, batch_size, current_key)
```
Comment (thomaspinder) [inline gpjax/abstractions.py:172]: I would prefer that the new batch was generated from a specific key i.e., the new key is pased into `get_batch_new_key()`, not the existing key being split within the function. I think for reproducibility, this approach will be cleaner. Unless there's a specific reason for structuring the code as you have done, then I would suggest changing to mimic my suggestion.

File: gpjax/abstractions.py [hunk 7]
```diff
@@ -162,33 +161,44 @@ def fit_batches(
...
+        train_data (Dataset): The training dataset.
+        batch_size (int): The batch size.
+    Returns:
+        Dataset: The batched dataset.
+    """
+    x, y, n = train_data.X, train_data.y, train_data.n
+
+    indicies = jr.choice(key, n, (batch_size,), replace=True)
+
+    _, new_key = jr.split(key)
```
Comment (thomaspinder) [inline gpjax/abstractions.py:202]: Re. comment on ln 172.

File: gpjax/types.py [hunk 8]
```diff
@@ -5,6 +5,8 @@
 
 NoneType = type(None)
 
+from chex import PRNGKey as PRNGKeyType
```
Comment (thomaspinder) [inline gpjax/types.py:8]: Unnecessary given line 2.

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Just two simple changes my side, then this can be merged.

File: gpjax/types.py (hunk 8 above)
Comment (Daniel-Dodd) [reply gpjax/types.py:8]: Oh yes. Good spot.

File: gpjax/abstractions.py (hunk 6 above)
Comment (Daniel-Dodd) [reply gpjax/abstractions.py:172]: Instead of having (in the function `get_batch_new_key`) the order:

```python
  indicies = jr.choice(key, n, (batch_size,), replace=True)

   _, new_key = jr.split(key)
```

You would prefer (?):
```python
   key, new_key = jr.split(key)
   indicies = jr.choice(key, n, (batch_size,), replace=True)
```

Is this what you mean?

Comment (Daniel-Dodd) [reply gpjax/abstractions.py:172]: (If so, I'd also prefer this).

Comment (thomaspinder) [reply gpjax/abstractions.py:172]: Yes. In fact, to be consistent with Jax documentation, I'd really like 
```python
 key, subkey = jr.split(key)
 indicies = jr.choice(subkey, n, (batch_size,), replace=True)
```

Comment (Daniel-Dodd) [reply gpjax/abstractions.py:172]: Fantastic! I'll get this done now.

Comment (Daniel-Dodd) [reply gpjax/abstractions.py:172]: Resolved in c31df38

**Comment by Daniel-Dodd** [comment]
> Just two simple changes my side, then this can be merged.

Hi @thomaspinder. These should be resolved in c31df38.

File: gpjax/abstractions.py [hunk 9]
```diff
@@ -162,33 +161,44 @@ def fit_batches(
...
+    """Batch the data into mini-batches.
+    Args:
+        train_data (Dataset): The training dataset.
+        batch_size (int): The batch size.
+    Returns:
+        Dataset: The batched dataset.
+    """
+    x, y, n = train_data.X, train_data.y, train_data.n
+
+    key, new_key = jr.split(key)
```
Comment (thomaspinder) [inline gpjax/abstractions.py:200]: Apologies. Maybe I wasn't very clear. My point was that I don't like the key being split with the function that collects my batch. If I run this function, then I want to be able to pass the function a key and then every time I run the function with the same key I will receive the same data batch. Right now, this won't happen as you're splitting the key within the function. In my opinion, the key should be split __within the training loop__ and then then the new key is passed to this function from which a data batch is returned. The function would then look like:
```python
def get_batch_new_key(
    train_data: Dataset, batch_size: int, key: PRNGKeyType
) -> Dataset:
    """Batch the data into mini-batches.
    Args:
        train_data (Dataset): The training dataset.
        batch_size (int): The batch size.
    Returns:
        Dataset: The batched dataset.
    """
    x, y, n = train_data.X, train_data.y, train_data.n
    indicies = jr.choice(key, n, (batch_size,), replace=True)
    return Dataset(X=x[indicies], y=y[indicies]), new_key
```

File: gpjax/abstractions.py (hunk 6 above)
Comment (thomaspinder) [inline gpjax/abstractions.py:172]: In accordance with my below comment, this line would become two:
```python
current_key, subkey = jr.split(current_key)
batch = get_batch_new_key(train_data, batch_size, subkey)
```

**Review by thomaspinder — APPROVED** [review]
